### ISSUE TITLE
error msg "`Add`/`AddAssign` (etc) is not implemented for the type `_`" poor for beginners

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#34297]
Reduced version of code encountered during a tutorial:

``` rust
fn main() {
    let x = 0;
    x += 1.0;
}
```

Compiler error message:

```
error: the trait bound `_: std::ops::AddAssign<_>` is not satisfied [--explain E0277]
 --> <anon>:3:5
3 |>     x += 1.0;
  |>     ^^^^^^^^
help: the following implementations were found:
help:   <f64 as std::ops::AddAssign>
help:   <f32 as std::ops::AddAssign>
help:   <u64 as std::ops::AddAssign>
help:   <i64 as std::ops::AddAssign>
help: and 21 others

error: aborting due to previous error
```

Namely, the error message says nothing about `i32` nor `f64` nor the use of `let x = 0;`, so a beginner really has no place to go for help here.

Can we leverage the `on_unimplemented` functionality to provide better feedback in the case where one is mixing `i32` and `f64`? (Or does it not have a way to put in a more specific message depending on the types involved in this case?)

There may be an issue for this elsewhere (we all know that our error messaging when type-inference goes to numeric-fallback), but the error message here was particularly bad, and I couldn't find an existing issue quickly, so I am filing this one to make sure this doesn't fall through the cracks.

**Comment by eddyb** [comment]
I think we can print `_` better, in cases where we don't have the expected/found long description.
`integer type: std::ops::AddAssign<floating-point type>` is already mildly better, but we can combine this with `#[on_unimplemented]` to get something like `floating-point type cannot be added to integer type`.

We can also sneak in the literal which created the inference variable, but in the current `ppaux` setup that's a bit problematic (would have to put a mapping in TLS just to print such things, or redo `ppaux`).

**Comment by sanxiyn** [comment]
After #35080, we now print `the trait bound {integer}: std::ops::AddAssign<{float}> is not satisfied`.

**Comment by DiThi** [comment]
As a beginner I stumbled upon this problem, but with integers, and a more complicated expression. The problem was much more clearly explained (what errors in the expression) when I replaced `x += ` by `x = x +`

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the trait bound `{integer}: std::ops::AddAssign<{float}>` is not satisfied
 --> src/main.rs:3:7
  |
3 |     x += 1.0;
  |       ^^ no implementation for `{integer} += {float}`
  |
  = help: the trait `std::ops::AddAssign<{float}>` is not implemented for `{integer}`
```

We could maybe add some suggestions to cast from float to integer or vice versa.

**Comment by estebank** [comment]
The casting suggestion could be added to `*Assign` traits' `rustc_on_unimplemented` attr once #44755 is implemented.

**Comment by estebank** [comment]
Current output:

```
error[E0277]: cannot add-assign `{float}` to `{integer}`
 --> src/main.rs:3:7
  |
3 |     x += 1.0;
  |       ^^ no implementation for `{integer} += {float}`
  |
  = help: the trait `std::ops::AddAssign<{float}>` is not implemented for `{integer}`
```

**Comment by estebank** [comment]
Closing in favor of #38564.
